src/util/threadpool.rs

```diff
@@ -280,7 +272,21 @@ impl<Q: ScheduleQueue<T>, T: Debug> TaskPool<Q, T> {
     }
 
     fn pop_task(&mut self) -> Option<Task<T>> {
-        self.task_queue.pop()
+        let task = self.task_queue.pop();
+        if task.is_none() {
+            // try fill queue when queue is empty.
+            self.try_fill_queue();
```

Comment:
Since the one who  is pushed first may be not the one popped first, and there may be  more than one task in the channel. The schedule algorithm is  according to the `ScheduleQueue` provided in initialization. @siddontang